src/dag.rs

```diff
@@ -125,4 +125,240 @@ where
 }
 
 #[cfg(test)]
-mod tests {}
+mod tests {
+    use super::*;
+    use petgraph::graph::DiGraph;
+    use std::collections::HashSet;
+
+    #[test]
+    fn test_build_dag_success() {
+        // Test: Can we build a dag?
+        let nodes: HashSet<i32> = [1, 2, 3, 4].iter().cloned().collect();
+        let edges: HashSet<(i32, i32, &str)> = [
+            (1, 2, "edge1"),
+            (2, 3, "edge2"),
+            (2, 4, "edge3"),
+        ].iter().cloned().collect();
+
+        let result = DiGraph::build_dag(nodes, edges);
+        assert!(result.is_ok());
+
+        let dag = result.unwrap();
+        assert_eq!(dag.node_count(), 4);
+        assert_eq!(dag.edge_count(), 3);
+    }
+
+    #[test]
+    fn test_build_dag_rejects_self_loop() {
+        // Test: Do we reject if an edge has the same parent and child?
+        let nodes: HashSet<i32> = [1, 2].iter().cloned().collect();
+        let edges: HashSet<(i32, i32, &str)> = [
+            (1, 1, "self_loop"), // Self-loop
+        ].iter().cloned().collect();
+
+        let result = DiGraph::build_dag(nodes, edges);
+        assert!(result.is_err());
+
+        match result.unwrap_err() {
+            Error::Cyclical => (),
+            _ => panic!("Expected Cyclical error"),
+        }
```

Comment:
@coderabbitai Please use:
```rust
        assert!(matches!(
            result.unwrap_err(),
            Error::Cyclical,
        ));
```
Apply `assert!(matches!()` to all error unit tests.